daxa/mission/erosita.py

```diff
@@ -0,0 +1,703 @@
+            field_test = [field in self._miss_poss_field_types or self._miss_poss_fields for field in fields]
+            if not all(field_test):
+                bad_fields = np.array(input_fields)[~np.array(field_test)]
+                raise ValueError("Some field names or field types {bf} are not associated with this mission, please"
+                        "choose from the following fields; {gf} or field types; {gft}".format(
+                        bf=",".join(bad_fields), gf=",".join(self._miss_poss_fields), gft=",".join(self._miss_poss_field_types)))
+            
+            else:
+                # If its got to this stage that means the field where input as a mix of field types and field names
+                # I did this when i was tired so sorry it looks a but weird?? i think it might look weird
+                # Selecting the field types from the input
+                field_types = [ft for ft in fields if ft in self._miss_poss_field_types]
+                # Turning them into their field names
+                fields_in_types = CalPV_info.loc[CalPV_info["Field_Type"].isin(field_types), "Field_Name"].tolist()
+                # Selecting the field names from the input
+                field_names = [ft for ft in fields if ft in self._miss_poss_fields]
+                updated_fields = fields_in_types + field_names
+        
+        # Removing the duplicates from updated_fields
+        updated_fields = list(set(updated_fields))
+
+        # Return the chosen fields 
+        return updated_fields
+    
+    def _inst_filtering(self, evlist_path: str):
+        """
+        Method to filter event lists for eROSITACalPV data based on instrument choice.
+        
+        :param Union[List[str], str] gd_tscopes: The names of telescopes from which the user wishes to INCLUDE data from.
+        """
+
+        insts = self.chosen_instruments
+        # Getting a string of the numbers of the telescope modules to add to the file name
+        insts_str = ''.join(sorted(re.findall(r'\d+', ''.join(insts))))
+
+        # Only filtering the eventlist if this combination of instruments hasn't been filtered for yet
+        if os.path.exists(evlist_path[:-5] + '_if_{}.fits'.format(insts_str)):
+            pass
+        else:
+            # Reading in the file
+            hdul = fits.open(evlist_path)
+
+            # Selecting the telescope module number column
+            data = hdul[1].data
+            t_col = data["TM_NR"]
+        
+            # Just makes sure we can iterate across inst(s), regardless of how many there are
+            if not isinstance(insts, list):
+                insts = [insts]
+            # Putting inst names into correct format to search in t_col for 
+            gd_insts = [int(re.sub('[^0-9]','', tscope)) for tscope in insts]
+            
+            # Getting the indexes of events with the chosen insts
+            gd_insts_indx = np.hstack([(t_col==i).nonzero()[0] for i in gd_insts])
```

Comment:
hstack is turning the list comprehension into an array so it can be used to index the data

the .nonzero() turns the (t_col==i) from an array of Bools into the indexes of t_col==i is True

Reply:
I think np.where might be a better way of doing this - list comprehensions are going to be sloooow here I would think.

Reply:
I timed it and its like 7 times faster to do np.where